File: AOC_2023/Day 16/16.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <set>
#include <unordered_map>

using namespace std;

class Solution
{
    public:
    vector<string> grid;
    unordered_map<char, pair<int,int>> dir = {{'N', {-1,0}}, {'S', {1,0}}, {'E', {0,1}}, {'W', {0,-1}}};
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            grid.push_back(line);
        }
    }

    bool inValid(pair<int,int> curr){
        return curr.first < 0 || curr.first >= grid.size() || curr.second < 0 || curr.second >= grid[0].size();
    }

    void dfs(pair<int,int> curr, char d, vector<vector<set<char>>>& visited){
        if(inValid(curr)) return;
        if(visited[curr.first][curr.second].count(d)) return;
        visited[curr.first][curr.second].insert(d);
        if(grid[curr.first][curr.second] == '.'){
            pair<int,int> next = {curr.first + dir[d].first, curr.second + dir[d].second};
            dfs(next, d, visited);
        }
        else if(grid[curr.first][curr.second] == '/'){
            if(d == 'N'){
                pair<int,int> next = {curr.first + dir['E'].first, curr.second + dir['E'].second};
                dfs(next, 'E', visited);
            }
            else if(d == 'S'){
                pair<int,int> next = {curr.first + dir['W'].first, curr.second + dir['W'].second};
                dfs(next, 'W', visited);
            }
            else if(d == 'E'){
                pair<int,int> next = {curr.first + dir['N'].first, curr.second + dir['N'].second};
                dfs(next, 'N', visited);
            }
            else if(d == 'W'){
                pair<int,int> next = {curr.first + dir['S'].first, curr.second + dir['S'].second};
                dfs(next, 'S', visited);
            }
        }
        else if(grid[curr.first][curr.second] == '\\'){
            if(d == 'N'){
                pair<int,int> next = {curr.first + dir['W'].first, curr.second + dir['W'].second};
                dfs(next, 'W', visited);
            }
            else if(d == 'S'){
                pair<int,int> next = {curr.first + dir['E'].first, curr.second + dir['E'].second};
                dfs(next, 'E', visited);
            }
            else if(d == 'E'){
                pair<int,int> next = {curr.first + dir['S'].first, curr.second + dir['S'].second};
                dfs(next, 'S', visited);
            }
            else if(d == 'W'){
                pair<int,int> next = {curr.first + dir['N'].first, curr.second + dir['N'].second};
                dfs(next, 'N', visited);
            }
        }
        else if(grid[curr.first][curr.second] == '|'){
            if(d == 'N' || d == 'S'){
                pair<int,int> next = {curr.first + dir[d].first, curr.second + dir[d].second};
                dfs(next, d, visited);
            }
            else if(d == 'E' || d == 'W'){
                pair<int,int> next1 = {curr.first + dir['N'].first, curr.second + dir['N'].second}, next2 = {curr.first + dir['S'].first, curr.second + dir['S'].second};
                dfs(next1, 'N', visited);
                dfs(next2, 'S', visited);
            }
        }
        else if(grid[curr.first][curr.second] == '-'){
            if(d == 'E' || d == 'W'){
                pair<int,int> next = {curr.first + dir[d].first, curr.second + dir[d].second};
                dfs(next, d, visited);
            }
            else if(d == 'N' || d == 'S'){
                pair<int,int> next1 = {curr.first + dir['E'].first, curr.second + dir['E'].second}, next2 = {curr.first + dir['W'].first, curr.second + dir['W'].second};
                dfs(next1, 'E', visited);
                dfs(next2, 'W', visited);
            }
        }  
    }

    int energise(pair<int,int> start, char dir){
        vector<vector<set<char>>> visited(grid.size(), vector<set<char>>(grid[0].size(), set<char>()));
        int result = 0;
        dfs(start, dir, visited);
        for(auto row: visited){
            for(auto col: row){
                if(col.size() > 0) result++;
            }
        }
        return result;
    }

    void part1(){
        int result = energise({0,0}, 'E');
        cout<<"Part 1: "<<result<<"\n";
    }

    void part2(){
        int result = 0;
        // leftmost and rightmost column
        for(int i = 0; i < grid.size();i++){
            result = max(result, energise({i,0}, 'E'));
            result = max(result, energise({i,grid[0].size()-1}, 'W'));
        }
        // topmost and bottommost row
        for(int i = 0; i < grid[0].size();i++){
            result = max(result, energise({0,i}, 'S'));
            result = max(result, energise({grid.size()-1,i}, 'N'));
        }
        cout<<"Part 2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2023/Day 16/16.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var directions map[rune][2]int =  map[rune][2]int{'N' : {-1, 0},'E' : {0, 1},'S' : {1, 0},'W' : {0, -1},}

type point struct {
	x int
	y int
}

func initialize(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return grid, nil
}

func energise(grid []string, start point, dir rune) int {
	visited := make(map[point]map[rune]bool)
	var dfs func(point, rune)
	dfs = func(p point, d rune) {
		if p.x < 0 || p.x >= len(grid) || p.y < 0 || p.y >= len(grid[0]){
			return
		}
		if _, ok := visited[p]; !ok {
			visited[p] = make(map[rune]bool)
		} else if _, ok := visited[p][d]; ok {
			return
		}
		visited[p][d] = true
		if grid[p.x][p.y] == '|' &&  strings.Contains("EW", string(d)) {
			dfs(point{p.x + directions['N'][0], p.y + directions['N'][1]}, 'N')
			dfs(point{p.x + directions['S'][0], p.y + directions['S'][1]}, 'S')
		} else if grid[p.x][p.y] == '-' && strings.Contains("NS", string(d)) {
			dfs(point{p.x + directions['E'][0], p.y + directions['E'][1]}, 'E')
			dfs(point{p.x + directions['W'][0], p.y + directions['W'][1]}, 'W')
		} else if grid[p.x][p.y] == '/' {
			if d == 'N' {
				dfs(point{p.x + directions['E'][0], p.y + directions['E'][1]}, 'E')
			} else if d == 'E' {
				dfs(point{p.x + directions['N'][0], p.y + directions['N'][1]}, 'N')
			} else if d == 'S' {
				dfs(point{p.x + directions['W'][0], p.y + directions['W'][1]}, 'W')
			} else if d == 'W' {
				dfs(point{p.x + directions['S'][0], p.y + directions['S'][1]}, 'S')
			}
		} else if grid[p.x][p.y] == '\\' {
			if d == 'N' {
				dfs(point{p.x + directions['W'][0], p.y + directions['W'][1]}, 'W')
			} else if d == 'E' {
				dfs(point{p.x + directions['S'][0], p.y + directions['S'][1]}, 'S')
			} else if d == 'S' {
				dfs(point{p.x + directions['E'][0], p.y + directions['E'][1]}, 'E')
			} else if d == 'W' {
				dfs(point{p.x + directions['N'][0], p.y + directions['N'][1]}, 'N')
			}
		} else{
			dfs(point{p.x + directions[d][0], p.y + directions[d][1]}, d)
		}
	}	
	dfs(start, dir)
	return len(visited)
}

func part1(grid []string) {
	fmt.Println("Part 1:", energise(grid, point{0, 0}, 'E'))
}

func part2(grid []string) {
	result := 0
	for i := 0; i < len(grid); i++ {
		result = max(result, energise(grid, point{i, 0}, 'E'))
		result = max(result, energise(grid, point{i, len(grid[0]) - 1}, 'W'))
	}
	for i := 0; i < len(grid[0]); i++ {
		result = max(result, energise(grid, point{0, i}, 'S'))
		result = max(result, energise(grid, point{len(grid) - 1, i}, 'N'))
	}
	fmt.Println("Part 2:", result)
}

func main(){
	filename := "input.txt"
	grid, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	part1(grid)
	part2(grid)
}